## Projeto 1 - SCC0270 - Redes Neurais e Aprendizado Profundo

### Nome: Nilo Conrado Messias Alves Cangerana
### Número USP: 9805362

In [1]:
#Imports
import platform
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import SGDClassifier

In [2]:
print("Versão Python utilizada: ",platform.python_version()) 

Versão Python utilizada:  3.7.9


In [2]:
#Carregar train.csv e test.csv
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')

In [4]:
#Checagem dos dados de treino
df_train.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V20       V21       V22       V23  \
0  0.098698  0.363787  0.090794  ...  0.251412 -0.018307  0.277838 -0.110474   
1  0.085102 -0.255425 -0.166974  ... -0.069083 -0.225775 -0.638672  0.101288   
2  0.247676 -1.514654  0.207643  ...  0.524980  0.247998  0.771679  0.909412   
3  0.377436 -1.387024 -0.054952  ... -0.208038 -0.108300  0.005274 -0.190321   
4 -0.270533  0.817739  0.753074  ...  0.408542 -0.009431  0.798278 -0.137458   

        V24       V25       V26       V27       V28  Class  
0  0.066928  0.128539 -0.189115  0.133558 -0.021053      0  
1 -0.339846  0.167170  0.125895 -0.008983  0.014724      0  
2 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752      0  
3 -1.175575  0.647376 -0.221929  0.062723  0.061458      0  
4  0.141267 -0.206010  0.502292  0.219422  0.215153      0  

[5 rows x 29 columns]

In [5]:
#Checagem dos dados de teste
df_test.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -0.439952  0.683758  1.225814  0.639113  0.716765  0.089295  0.657718   
1 -4.868108  1.264420 -5.167885  3.193648 -3.045621 -2.096166 -6.445610   
2  1.013114 -0.334412  1.305208  0.837406 -1.126833 -0.064321 -0.594753   
3  0.969231 -0.233554  0.238473  0.145793 -0.545741 -0.970680  0.347393   
4 -0.856523  1.080875  1.866956  1.729941 -0.161741  0.028789  0.401787   

         V8        V9       V10  ...       V20       V21       V22       V23  \
0  0.034213 -0.472679  0.352698  ...  0.211867  0.127443  0.588590 -0.145728   
1  2.422536 -3.214055 -8.745973  ...  0.667310  1.269205  0.057657  0.629307   
2  0.147737  0.536360 -0.120472  ...  0.069161  0.017079  0.112210 -0.016084   
3 -0.209522 -0.342571 -0.100331  ...  0.240978 -0.362820 -1.417272  0.162136   
4  0.043774 -0.213916  0.155907  ... -0.068915  0.007365  0.077392 -0.221906   

        V24       V25       V26       V27       V28  Class  
0 -0.322900 -0.299744 -0.310295  0.248317 -0.011493      0  
1 -0.168432  0.443744  0.276539  1.441274 -0.127944      1  
2  0.595033  0.201073  0.278215  0.007457  0.030762      0  
3  0.541628 -0.079465  0.268702 -0.101237  0.028234      0  
4  0.394141  0.237225 -0.080102 -0.291408  0.092140      0  

[5 rows x 29 columns]

In [6]:
#Checagem do shape 
print('Shape train:',df_train.shape)
print('Shape test:',df_test.shape)

Shape train: (142404, 29)
Shape test: (142403, 29)


In [3]:
#Gerar X_train, y_train, X_test, y_test em numpy array
X_train=df_train.to_numpy()
X_test=df_test.to_numpy()

y_train=X_train[:, X_train.shape[1]-1]
y_test=X_test[:, X_test.shape[1]-1]

X_train=X_train[:,:-1]
X_test=X_test[:,:-1]

In [4]:
#Checagem dos shapes:
print('Shape X_train:',X_train.shape)
print('Shape y_train:',y_train.shape,'\n')

print('Shape X_test:',X_test.shape)
print('Shape y_test:',y_test.shape)

Shape X_train: (142404, 28)
Shape y_train: (142404,) 

Shape X_test: (142403, 28)
Shape y_test: (142403,)


## Questão 1

In [9]:
#Contar quantidade de fraudes nos dados
qtdFraudesTrain = np.count_nonzero(y_train == 1)
qtdFraudesTest = np.count_nonzero(y_test == 1)

print('Quantidade de fraudes(classe 1) no conjunto de treino:',qtdFraudesTrain,'de',len(X_train),'amostras total de treino.')
print('Quantidade de não fraudes(classe 0) no conjunto de treino:',len(X_train)-qtdFraudesTrain,'de',len(X_train),'amostras total de treino.\n')
print('Quantidade de fraudes(classe 1) no conjunto de teste:',qtdFraudesTest,'de',len(X_test),'amostras total de teste.')
print('Quantidade de não fraudes(classe 0) no conjunto de teste:',len(X_test)-qtdFraudesTest,'de',len(X_test),'amostras total de teste.')
qtdNaoFraudesTest=len(X_test)-qtdFraudesTest

Quantidade de fraudes(classe 1) no conjunto de treino: 269 de 142404 amostras total de treino.
Quantidade de não fraudes(classe 0) no conjunto de treino: 142135 de 142404 amostras total de treino.

Quantidade de fraudes(classe 1) no conjunto de teste: 223 de 142403 amostras total de teste.
Quantidade de não fraudes(classe 0) no conjunto de teste: 142180 de 142403 amostras total de teste.


<b>a)</b> A métrica de avaliação que deve ser utilizada para este problema é a AUC. Como é possível ver nos dados, a quantidade de fraudes é muito inferior a quantidade de transações verdadeiras e, portanto, um modelo utilizando a métrica de acurácia poderia resultar em uma acurácia muito alta, mesmo errando todas previsões de fraudes, tornando o modelo ineficiente para seu propósito.

Além disso, os dados possuem classes desbalanceadas, dado que existem menos fraudes nos conjuntos, e é necessário que o modelo acerte o máximo de fraudes possíveis. Para isso, o modelo que tiver a melhor AUC,ou seja, melhor custo beneficio na curva ROC(mais verdadeiros positivos e menos falsos positivos) será o melhor.

<b>b)</b> Acurácia: Como o modelo é aleatório, sabe-se que a probabilidade do modelo prever classe 0 é 0.5 e a probabilidade do modelo prever classe 1 é 0.5. 

In [10]:
print('(P1)Probabilidade do dado ser classe 0 e classificado como classe 0:',(qtdNaoFraudesTest/len(X_test))*0.5)
print('(P2)Probabilidade do dado ser classe 1 e classificado como classe 1:',(qtdFraudesTest/len(X_test))*0.5)
print('Portanto, a acurácia do modelo aleatório é P1+P2=',(qtdNaoFraudesTest/len(X_test))*0.5+(qtdFraudesTest/len(X_test))*0.5)

(P1)Probabilidade do dado ser classe 0 e classificado como classe 0: 0.4992170108775798
(P2)Probabilidade do dado ser classe 1 e classificado como classe 1: 0.0007829891224201737
Portanto, a acurácia do modelo aleatório é P1+P2= 0.5


AUC: Quando o modelo tem a mesma probabilidade de classificar aleatóriamente um dado, a curva ROC do modelo é uma linha diagonal que vai do ponto (0,0) até (1,1). Neste caso, a área sobre essa curva é a área de um triângulo que é igual a (base*altura)/2

Portanto, a AUC do modelo aleatório é: (1*1)/2=0.5

## Questão 2

In [11]:
#Instanciar MLPCLassifier para cada valor de alpha
m1=MLPClassifier(hidden_layer_sizes = (20,20,20,20,20,), activation='relu',alpha=0.0 ,random_state=42)
m2=MLPClassifier(hidden_layer_sizes = (20,20,20,20,20,), activation='relu',alpha=0.01 ,random_state=42)
m3=MLPClassifier(hidden_layer_sizes = (20,20,20,20,20,), activation='relu',alpha=0.1 ,random_state=42)
m4=MLPClassifier(hidden_layer_sizes = (20,20,20,20,20,), activation='relu',alpha=1 ,random_state=42)
m5=MLPClassifier(hidden_layer_sizes = (20,20,20,20,20,), activation='relu',alpha=10 ,random_state=42)

In [12]:
#Treinar modelo para alpha=0.0
m1.fit(X_train,y_train)

MLPClassifier(alpha=0.0, hidden_layer_sizes=(20, 20, 20, 20, 20),
              random_state=42)

In [13]:
#Treinar modelo para alpha=0.01
m2.fit(X_train,y_train)

MLPClassifier(alpha=0.01, hidden_layer_sizes=(20, 20, 20, 20, 20),
              random_state=42)

In [14]:
#Treinar modelo para alpha=0.1
m3.fit(X_train,y_train)

MLPClassifier(alpha=0.1, hidden_layer_sizes=(20, 20, 20, 20, 20),
              random_state=42)

In [15]:
#Treinar modelo para alpha=1
m4.fit(X_train,y_train)

MLPClassifier(alpha=1, hidden_layer_sizes=(20, 20, 20, 20, 20), random_state=42)

In [16]:
#Treinar modelo para alpha=10
m5.fit(X_train,y_train)

MLPClassifier(alpha=10, hidden_layer_sizes=(20, 20, 20, 20, 20),
              random_state=42)

In [17]:
#Medir AUC de cada modelo
m1_AUC=roc_auc_score(y_test,m1.predict(X_test))
m2_AUC=roc_auc_score(y_test,m2.predict(X_test))
m3_AUC=roc_auc_score(y_test,m3.predict(X_test))
m4_AUC=roc_auc_score(y_test,m4.predict(X_test))
m5_AUC=roc_auc_score(y_test,m5.predict(X_test))

print('Taxa de Regularização L2=0.0 - AUC=', m1_AUC)
print('Taxa de Regularização L2=0.01 - AUC=', m2_AUC)
print('Taxa de Regularização L2=0.1 - AUC=', m3_AUC)
print('Taxa de Regularização L2=1 - AUC=', m4_AUC)
print('Taxa de Regularização L2=10 - AUC=', m5_AUC)

Taxa de Regularização L2=0.0 - AUC= 0.8890056783954149
Taxa de Regularização L2=0.01 - AUC= 0.8896808788455485
Taxa de Regularização L2=0.1 - AUC= 0.8922043648328054
Taxa de Regularização L2=1 - AUC= 0.600896860986547
Taxa de Regularização L2=10 - AUC= 0.5


Como é possível ver nas métricas medidas acima, um modelo com taxa de regularização igual a 0 tem menos acertos em relação a um modelo com taxas de regularização maiores(0.01 e 0.1). Isso se deve ao fato do modelo com regularização reduzir sua complexidade para evitar um possível overfitting. Para taxas de regularização maiores(1 e 10), o fator da regularização na fórmula do erro tem um peso maior, fazendo com que não compense para o modelo usar toda sua complexidade para reduzir o erro. Isso torna o modelo incapaz de generalizar com toda sua complexidade, diminuindo drasticamente o valor da AUC para esses casos.

Portanto, para uma taxa de regularização igual a 0.1, o modelo possui a melhor performance.

## Questão 3

In [18]:
#Instanciar MLPClassifier
modelMLP = MLPClassifier(random_state=42)

#Instanciar KNN
modelKNN = KNeighborsClassifier()

#Definição dos hiperparametros que serão variados em cada modelo:
hyperMLP = {'hidden_layer_sizes':[(), (10,), (5,5,)]}
hyperKNN = {'n_neighbors':[3, 5, 7]}

In [19]:
#Instanciar GridSearchCV para MLP
gsMLP=GridSearchCV(modelMLP, hyperMLP, scoring='roc_auc', cv=3, verbose=2)

#Treinar MLP
gsMLP.fit(X_train,y_train)

Fitting 3 folds for each of 3 candidates, totalling 9 fits
[CV] END ..............................hidden_layer_sizes=(); total time=   4.5s
[CV] END ..............................hidden_layer_sizes=(); total time=   5.4s
[CV] END ..............................hidden_layer_sizes=(); total time=   4.8s
[CV] END ...........................hidden_layer_sizes=(10,); total time=   3.8s
[CV] END ...........................hidden_layer_sizes=(10,); total time=   5.2s
[CV] END ...........................hidden_layer_sizes=(10,); total time=   5.2s
[CV] END ..........................hidden_layer_sizes=(5, 5); total time=   5.0s
[CV] END ..........................hidden_layer_sizes=(5, 5); total time=   5.0s
[CV] END ..........................hidden_layer_sizes=(5, 5); total time=   6.6s


GridSearchCV(cv=3, estimator=MLPClassifier(random_state=42),
             param_grid={'hidden_layer_sizes': [(), (10,), (5, 5)]},
             scoring='roc_auc', verbose=2)

In [20]:
print('Melhor configuração para MLP:',gsMLP.best_estimator_)
print('AUC da melhor configuração:',gsMLP.best_score_)

Melhor configuração para MLP: MLPClassifier(hidden_layer_sizes=(10,), random_state=42)
AUC da melhor configuração: 0.9544289035505876


O MLPClassifier com uma camada intermediária de 10 neurônios apresenta a melhor AUC.

In [21]:
#Instanciar GridSearchCV para KNN
gsKNN=GridSearchCV(modelKNN, hyperKNN, scoring='roc_auc',cv=3, verbose=2)

#Treinar KNN
gsKNN.fit(X_train,y_train)

Fitting 3 folds for each of 3 candidates, totalling 9 fits
[CV] END ......................................n_neighbors=3; total time= 1.1min
[CV] END ......................................n_neighbors=3; total time= 1.2min
[CV] END ......................................n_neighbors=3; total time= 1.2min
[CV] END ......................................n_neighbors=5; total time= 1.5min
[CV] END ......................................n_neighbors=5; total time= 1.5min
[CV] END ......................................n_neighbors=5; total time= 1.6min
[CV] END ......................................n_neighbors=7; total time= 1.5min
[CV] END ......................................n_neighbors=7; total time= 1.5min
[CV] END ......................................n_neighbors=7; total time= 1.6min


GridSearchCV(cv=3, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7]}, scoring='roc_auc',
             verbose=2)

In [22]:
print('Melhor configuração para KNN:',gsKNN.best_estimator_)
print('AUC da melhor configuração:',gsKNN.best_score_)

Melhor configuração para KNN: KNeighborsClassifier(n_neighbors=7)
AUC da melhor configuração: 0.91067105000196


O KNN com n_neighbors=7 apresenta a melhor AUC.

In [23]:
%%time
#Previsão na base de teste para MLP
AUC_bestMLP=roc_auc_score(y_test,gsMLP.best_estimator_.predict(X_test))

Wall time: 75.8 ms


In [24]:
%%time
#Previsão na base de teste para KNN
AUC_bestKNN=roc_auc_score(y_test,gsKNN.best_estimator_.predict(X_test))

Wall time: 7min 15s


In [25]:
print('AUC melhor MLP:',AUC_bestMLP)
print('AUC melhor KNN:',AUC_bestKNN)

AUC melhor MLP: 0.8829192547563343
AUC melhor KNN: 0.8562736586667441


<b>a)</b> A técnica que obteve o melhor resultado é o modelo de rede neural com MLPClassifier, com AUC de aproximadamente 0.883

<b>b)</b> A técnina que demora mais para gerar resultados na base de teste é a KNN, conforme estimado acima pelo comando %%time

## Questão 4

In [26]:
fraudeBloq=100
naoFraudeBloq=-2

#Probabilidades de fraudes no conjunto de teste
predictProba=gsMLP.best_estimator_.predict_proba(X_test)

In [27]:
predictProba[:10]

array([[9.99928171e-01, 7.18291665e-05],
       [4.24865807e-02, 9.57513419e-01],
       [9.99932555e-01, 6.74452913e-05],
       [9.99999808e-01, 1.91914416e-07],
       [9.99916173e-01, 8.38272725e-05],
       [9.99999978e-01, 2.20523181e-08],
       [9.99972709e-01, 2.72911055e-05],
       [9.99926227e-01, 7.37734496e-05],
       [9.99999999e-01, 7.83388495e-10],
       [9.99999215e-01, 7.84995728e-07]])

In [28]:
#Ordenar vetor por probabilidade de ser fraude. Da maior probabilidade até a menor:
predictProbaSorted=sorted(enumerate(predictProba[:,1]), key=lambda i: i[1], reverse=True)

#Formato da lista gerada: (Indice do valor no array predictProba, Probabilidade de ser fraude)
predictProbaSorted[:10]

[(101345, 0.999995902524908),
 (101295, 0.999994973675985),
 (101444, 0.9999922335068034),
 (11419, 0.9999921775299898),
 (101600, 0.9999852631074735),
 (101929, 0.9999720108183268),
 (8602, 0.9999589500513425),
 (8603, 0.9999589500513425),
 (8604, 0.9999589500513425),
 (8605, 0.9999589500513425)]

In [29]:
#Gerando uma lista com apenas o top 1% das transações com mais chances de ser fraude
#top 1%: 1424 amostras
topFraudes=predictProbaSorted[:int(len(predictProbaSorted)*0.01)]

print('Quantidade de amostras(1% de 142403):',len(topFraudes))

Quantidade de amostras(1% de 142403): 1424


In [30]:
topFraudes[:10]

[(101345, 0.999995902524908),
 (101295, 0.999994973675985),
 (101444, 0.9999922335068034),
 (11419, 0.9999921775299898),
 (101600, 0.9999852631074735),
 (101929, 0.9999720108183268),
 (8602, 0.9999589500513425),
 (8603, 0.9999589500513425),
 (8604, 0.9999589500513425),
 (8605, 0.9999589500513425)]

De acordo com o enunciado, todas as 1424 transações seriam impedidas de acontecer, ou seja, seriam bloqueadas.

In [44]:
#Calcular lucro
lucro=0
for i in topFraudes:
    if y_test[i[0]]==1: #se um transação i for fraude e foi impedida de acontecer
        lucro+=fraudeBloq
    elif y_test[i[0]]==0: #se uma transação i não for fraude e foi impedida de acontecer
        lucro+=naoFraudeBloq

print('Lucro calculado:',lucro)

Lucro calculado: 16430


O lucro calculado para este modelo é de R$ 16430

## Questão 5

In [5]:
#Instanciar modelo de SGDClassifier
modelSGD=SGDClassifier(loss='log', penalty='l1', alpha=0.00003, random_state=42)

Foram testados vários valores para alpha, para atingir o melhor AUC. O melhor AUC para o modelo foi para alpha=0.00003

In [6]:
#Treinar modelo
modelSGD.fit(X_train,y_train)

SGDClassifier(alpha=3e-05, loss='log', penalty='l1', random_state=42)

In [7]:
#Calcular AUC
sgdAUC=roc_auc_score(y_test,modelSGD.predict(X_test))
print('AUC do modelo SGDClassifier',sgdAUC)

AUC do modelo SGDClassifier 0.8342013881223006


In [8]:
#Pesos atribuidos a cada entrada:
modelSGD.coef_

array([[ 131.20895405,  154.55633083,    6.85731862,   22.85853157,
         -47.10080766,    0.        ,  -13.07348791,   12.29360645,
         -82.40875519,   -0.51276085,  -48.06424068,   36.61012713,
         -87.48998568,  -78.73042921,  -55.76757215,   -8.26104004,
         -14.34832002,    0.        ,  -15.03392036,   63.19105309,
          61.90761457,   42.33356232,   20.3493302 ,   -2.54213316,
         -65.58029015,   15.61497523, -116.79512123,   20.05153403]])

As variáveis não importantes para o problema são aquelas em que foram atribuídos pesos zero.

In [19]:
index=1
listVar=[]
for i in modelSGD.coef_[0]:
    if i==0:
        listVar.append('V'+str(index))
    index+=1
print('Variáveis não importantes:')
print(listVar)

Variáveis não importantes:
['V6', 'V18']


Essas variáveis não influenciam na previsão final do modelo, dado que a regularização L1 faz com que seus pesos sejam zerados, reduzindo a complexidade do modelo.